### ISSUE TITLE
De-`unwrap` all assists

### ISSUE THREAD

**Issue by lnicola** [rust-lang/rust-analyzer#15398]
Historically, the assists have been one of our major source of crashes, usually because they didn't handle invalid or incomplete code. Let's try to clean these up a bit.

Here are the ones I've found, though some of the usages there might actually be correct.

- [ ] generate_derive
- [ ] generate_delegate_trait #15397
- [ ] remove_unused_imports
- [ ] inline_call
- [ ] extract_function
- [ ] add_missing_match_arms
- [ ] wrap_return_type_in_result
- [ ] generate_delegate_methods
- [ ] desugar_doc_comment
- [ ] remove_dbg
- [ ] generate_default_from_new
- [ ] convert_to_guarded_return
- [ ] convert_comment_block
- [ ] convert_named_struct_to_tuple_struct
- [ ] apply_demorgan

**Comment by lowr** [comment]
The [`unwrap()` in `remove_dbg`](https://github.com/rust-lang/rust-analyzer/blob/70fa270637193a4eb35fc7c07728c4e1f79a118b/crates/ide-assists/src/handlers/remove_dbg.rs#L49) is fine because `replacements` is checked to be non-empty right before it.

**Comment by lnicola** [comment]
That's right, but we could still extract that into a local variable and check it properly. It's not worse, and we won't spend time figuring it out again the next 10 times we go through the assists.

**Comment by lowr** [comment]
A comment would suffice I'd say. It's definitely bad to `unwrap()` without any explanation but for those as simple as this one I'd rather not write unnecessary control flow. I'll file a PR that adds a comment if that sounds reasonable.

Meanwhile, I've found that `unwrap()` in `generate_default_from_new` *is* problematic, sigh.

**Comment by lnicola** [comment]
I mean, a `?` (or even my original suggestion) will let us remove the check above, so it's not more control flow. It's much easier to grep and not have the unwrap there than to read the comment and think 😅.

**Comment by lnicola** [comment]
By the way, my idea here is to clean up as much of these as we can, then add a "ratchet"-style test to make sure we're not adding more by mistake.

**Comment by Milo123459** [comment]
can this be closed because all the tasks have been completed?

**Comment by lnicola** [comment]
Almost. I think there are a couple more in `generate_function` which we can turn into `expect`.

**Comment by Milo123459** [comment]
can i potentially look into doing those?

**Comment by lnicola** [comment]
Sure. Check out lines 407 and 686 in `generate_function.rs`.

The other thing I wanted to have is a `tidy` test to count how many unwraps there are in this directory.

**Comment by Milo123459** [comment]
is there a place i can ask for guidance? some of these aren't that clear - for example line 406 suggests that the unwrap is fine? or am i misinterpreting it.

**Comment by lnicola** [comment]
Yeah, I meant replacing `.unwrap()` with `.expect("generated function should have a body")` and so on.

**Comment by Milo123459** [comment]
should be all good, not sure about the `tail_expr` expect.
